Add VPP support

AloysAugustin opened the pull request:
## Description

This PR adds support for VPP in the operator. We hope that it will enable the operator to become the primary way of managing Calico installations with the VPP dataplane enabled.
It should have no impact when VPP-specific configuration is not added. It changes the Installation object to add `VPP` as an option for `linuxDataplane`, and adds all the VPP-specific options in a `vpp` object under the `CalicoNetwork` object. 

It has been tested manually on VM, baremetal and EKS clusters, which are the 3 deployments currently officially supported by VPP.

<!-- A few sentences describing the overall goals of the pull request's commits.
Please include
- the type of fix - (e.g. bug fix, new feature, documentation)
- some details on _why_ this PR should be merged
- the details of the testing you've done on it (both manual and automated)
- which components are affected by this PR
- links to issues that this PR addresses
-->

## For PR author

- [ ] Tests for change.
- [x] If changing pkg/apis/, run `make gen-files`
- [x] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/tigera/operator?pullRequest=1415) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/tigera/operator?pullRequest=1415) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/tigera/operator?pullRequest=1415) it.</sub>

caseydavenport commented on pkg/render/node.go:
```diff
@@ -1484,11 +1893,20 @@ func GetIPv6Pool(pools []operatorv1.IPPool) *operatorv1.IPPool {
 	return instance.CalicoNetwork != nil &&
 		instance.CalicoNetwork.BGP != nil &&
-		*instance.CalicoNetwork.BGP == operatorv1.BGPEnabled
+		*instance.CalicoNetwork.BGP == operatorv1.BGPEnabled &&
```
I think this change is a bit confusing - instance.CalicoNetwork.BGP is meant to indicate whether or not BGP should be enabled - since VPP bundles its own BGP daemon, BGP is actually enabled, it's just not using BIRD.

Maybe it's as simple as renaming this function to `birdEnabled`? 

We probably also want to update validation to ensure that if VPP is enabled, instance.CalicoNetwork.BGP is also enabled - unless VPP data plane also supports running without BGP?

caseydavenport commented on pkg/render/node.go:
```diff
@@ -1484,11 +1893,20 @@ func GetIPv6Pool(pools []operatorv1.IPPool) *operatorv1.IPPool {
+func vppDataplaneEnabled(cr *operatorv1.InstallationSpec) bool {
+	return cr.CalicoNetwork != nil &&
+		cr.CalicoNetwork.LinuxDataplane != nil &&
+		*cr.CalicoNetwork.LinuxDataplane == operatorv1.LinuxDataplaneVPP &&
```
It's a bit of a red flag to me that we're relying on two API configs to determine if vpp is enabled. the LinuxDataplane option should be the deciding factor as to whether or not the VPP dataplane is being used.

The cr.CalicoNetwork.VPP should probably be defaulted based on the LinuxDataplane option, and validation should be used to make sure that the two are consistent with each other before ever reaching this function.

caseydavenport replied:
In other words - cr.CalicoNetwork.VPP should never be nil at this point in the code if LinuxDataplane == VPP

caseydavenport commented on pkg/render/node.go:
```diff
@@ -706,6 +798,46 @@ func (c *nodeComponent) nodeDaemonset(cniCfgMap *corev1.ConfigMap) *appsv1.Daemo
 }
 
+// vppDaemonset creates the vpp daemonset.
+func (c *nodeComponent) vppDaemonset() *appsv1.DaemonSet {
```
I think there is probably a case to made for this DaemonSet and supporting functions being in its own rendered - it's not really part of calico/node. @tmjd WDYT?

caseydavenport commented on pkg/render/node.go:
```diff
@@ -255,137 +289,180 @@ func (c *nodeComponent) nodeRoleBinding() *rbacv1.ClusterRoleBinding {
+// nodeRole creates the clusterrole containing policy rules that allow the node daemonset to operate normally.
+func (c *nodeComponent) nodeRole(forVpp bool) *rbacv1.ClusterRole {
+	var role *rbacv1.ClusterRole
+	if !forVpp {
```
I'm not a fan of this approach for re-using the ClusterRole struct between two different cluster role. I think we should just have two separate cluster roles.

caseydavenport commented on pkg/render/node.go:
```diff
@@ -143,6 +147,20 @@ func (c *nodeComponent) ResolveImages(is *operatorv1.ImageSet) error {
+	if vppDataplaneEnabled(c.cfg.Installation) {
+		c.vppImage, err = components.GetReference(components.ComponentCalicoVpp, reg, path, prefix, is)
+	}
+	if err != nil {
```
This error check should be within the `vppDataplaneEnabled` branch above where the error is generated

caseydavenport commented on pkg/render/node.go:
```diff
@@ -143,6 +147,20 @@ func (c *nodeComponent) ResolveImages(is *operatorv1.ImageSet) error {
+	if vppDataplaneEnabled(c.cfg.Installation) {
+		c.vppAgentImage, err = components.GetReference(components.ComponentCalicoVppAgent, reg, path, prefix, is)
+	}
+	if err != nil {
```
same for this guy

caseydavenport commented on api/v1/installation_types.go:
```diff
@@ -345,6 +350,76 @@ type CalicoNetworkSpec struct {
+	// IPsec can be used to enable IPsec encryption of IPIP tunnels
+	// Accepted values: "Disabled" (default), "Enabled"
+	// +optional
+	IPsec string `json:"ipsec,omitempty"`
```
Similar to SerivcePrefix, wondering if this should ultimately be outside of this struct and enforced using validation. In theory, other dataplanes could add IPSec support in the future.

caseydavenport reviewed (changes requested):
@AloysAugustin really sorry for the delay on this! To expedite, I think we should do what we talked about on the call and do this in two parts. I think it will also help clarify some of the logic for me as well.

1. Add the new LinuxDataplaneVPP option, which configures calico/node and CNI to use the external dataplane. The VPP daemonset itself can be deployed out-of-band.
2. Add the VPPConfiguration fields in a second PR, allowing deployment of VPP itself via the operator.

Hope that's not too much of a burden. 

I did review the VPPConfiguration section - I am still chewing on some of the details. Most of my comments were around aligning with [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md). 

The other area I'm thinking about is validation - I don't see much of it in this PR as it stands, but we definitely will need some - especially doing cross-field validation that the entire Installation is consistent. The pattern we follow in the operator is that the render code shouldn't have to make decisions for the user - the defaulting and validation code should ensure that an actionable, self-consistent configuration is passed to render. 

As an example, if VPP is enabled but BGP is disabled, what is the user asking for? It's not clear, so we should validate against it.

AloysAugustin commented on api/v1/installation_types.go:
```diff
@@ -345,6 +350,76 @@ type CalicoNetworkSpec struct {
+	// DefaultGateway can be set to configure a default route in VPP.
+	// Accepted formats are "ipv4", "ipv6" or "ipv4,ipv6"
+	// +optional
+	DefaultGateway string `json:"defaultGateway,omitempty"`
```
We actually expect one or two IP addresses, looks like the comment is misleading :)

AloysAugustin commented on api/v1/installation_types.go:
```diff
@@ -345,6 +350,76 @@ type CalicoNetworkSpec struct {
+
+	// HookType is set to specify node distribution, only "systemctl" is supported for now
+	// +optional
+	HookType string `json:"hookType,omitempty"`
```
It can also be empty - but like a few other fields, we should make it a proper enum, and default to something like "None"

AloysAugustin commented:
Thanks @caseydavenport for the review! I simplified this PR to include only the changes to calico-node as discussed. We'll open a new PR for the VPP specific bits, while taking your comments into account.

I've already added some level of config validation in this PR to ensure that the config is consistent (if VPP is enabled, the CNI should be calico and BGP should be enabled), and a few tests as well.

caseydavenport commented on pkg/render/node.go:
```diff
@@ -1220,6 +1245,27 @@ func (c *nodeComponent) nodeEnvVars() []corev1.EnvVar {
+				Name:  "FELIX_AWSSRCDSTCHECK",
+				Value: "Disable",
+			}, corev1.EnvVar{
+				Name:  "IP_AUTODETECTION_METHOD",
```
IP_AUTODETECTION_METHOD is configurable via the Installation API, so we can't ignore whatever the user has specified there. If VPP requires this, then we should set this in defaulting and validation rather than deep in the render code so that the Installation API is consistent with the rendered objects.

caseydavenport commented on pkg/render/node.go:
```diff
@@ -1220,6 +1245,27 @@ func (c *nodeComponent) nodeEnvVars() []corev1.EnvVar {
+		})
+		if c.cfg.Installation.KubernetesProvider == operatorv1.ProviderEKS {
+			nodeEnv = append(nodeEnv, corev1.EnvVar{
+				Name:  "FELIX_AWSSRCDSTCHECK",
```
Hm, I was expecting us to already be setting this but it doesn't look like it. @tmjd do you recall whether we discussed this in the past?

tmjd replied:
I don't remember talking about that one before.
I'd be concerned to default it because of this from the [reference docs](https://projectcalico.docs.tigera.io/reference/felix/configuration) 
> Check IAM role and profile configuration for setting the necessary permission for this setting to work. [Default: DoNothing]

AloysAugustin replied:
For this one, we've had a lot fewer issues by enabling it on EKS, and EKS nodes have the required IAM policy by default (as stated on [this page](https://projectcalico.docs.tigera.io/reference/resources/felixconfig#aws-iam-rolepolicy-for-source-destination-check-configuration)). So it would be nice to have a way to enable it. Would you prefer to expose this variable in the Installation spec and default it to `Disable` when using EKS + VPP?

caseydavenport replied:
I think it's OK to leave this as it is here - if we ever do add a config option for it, we can default that option based on existing behavior.

The only consideration would be if we want to set this via `FelixConfiguration` datastore API instead of an env variable - I believe that's the path we want to be going but I can't recall if we have the right render infrastructure set up to make that easy to do.

tmjd replied:
Sounds good to me to leave this.
As discussed in the Calico Community meeting we don't have the proper infrastructure for the FelixConfiguration to allow/understand if the users overwrote/set the field.

caseydavenport commented on pkg/render/node.go:
```diff
@@ -1254,7 +1300,11 @@ func (c *nodeComponent) nodeEnvVars() []corev1.EnvVar {
 		// BGP is enabled.
-		nodeEnv = append(nodeEnv, corev1.EnvVar{Name: "CALICO_NETWORKING_BACKEND", Value: "bird"})
+		if c.vppDataplaneEnabled() {
+			nodeEnv = append(nodeEnv, corev1.EnvVar{Name: "CALICO_NETWORKING_BACKEND", Value: "none"})
```
Probably worth a comment here saying that BGP is enabled, but that the VPP data plane has its own BGP implementation and so we disabled BIRD anyway. Otherwise this code might be not-obvious for future readers.

caseydavenport commented on pkg/controller/installation/validation.go:
```diff
@@ -247,6 +247,16 @@ func validateCustomResource(instance *operatorv1.Installation) error {
 		}
 
+		// VPP specific validation
+		if instance.Spec.CalicoNetwork.LinuxDataplane != nil && *instance.Spec.CalicoNetwork.LinuxDataplane == operatorv1.LinuxDataplaneVPP {
```
We should probably also verify that the `spec.Variant` field is set to `Calico` since we don't support this for Calico Enterprise just yet. 

@tmjd any other validations you think we should be checking here?

tmjd replied:
Are there any IPv4 and IPv6 limitations?
Any encapsulation limitations?
Are there any HostPorts or ContainerIPForwarding limitations?
I don't really know what limitations there would be, I was just looking at the fields in the Installation resource.

AloysAugustin replied:
Thanks, good questions :) 
IPv4 / IPv6 / dual stack is supported.
In terms of encap, we support flat IP / IPIP / VXLAN and wireguard so I think it should be alright.
HostPorts have been added recently, and ContainerIPForwarding should be supported but I don't think we've tested it, I'll make a note to check it.

caseydavenport replied:
> HostPorts have been added recently,

Perhaps a nitty point, but does VPP support _disabling_ hostports? e.g., if `hostPorts: Disabled` is set, will VPP respect that? If not, we need to validate that hostPorts is never set to Disabled while in VPP mode.

caseydavenport replied:
@AloysAugustin just one more question above, otherwise I think this is good to go

AloysAugustin replied:
We don't support disabling `HostPorts`, so I added a validation check for that.

caseydavenport commented:
@AloysAugustin thanks for reworking this guy - it's looking very close! I just added a couple thoughts.

AloysAugustin commented:
Thanks @caseydavenport  and @tmjd for the review, most comments should be addressed by the latest commit, looks like the last sticking point is the FELIX_AWSSRCDSTCHECK discussion.

caseydavenport commented:
@tmjd CI looks to be failing with a CRD issue - does this just need rebase to fix?

tmjd commented:
Yes, I merged a PR that should fix that earlier in the day.

AloysAugustin commented:
Thanks, I just rebased the branch.

AloysAugustin commented:
Looks like the CI failed again but I can't see the logs, any chance someone could take a look?

caseydavenport commented:
@AloysAugustin was just a test flake - all good now! :rocket:
